src/common/resources/request.rs

```diff
@@ -0,0 +1,176 @@
+use serde::{Deserialize, Serialize};
+
+use crate::common::error::DsError;
+use crate::common::resources::{GenericResourceAmount, GenericResourceId, NumOfCpus};
+use crate::worker::pool::ResourcePool;
+use smallvec::SmallVec;
+use std::time::Duration;
+
+#[derive(Serialize, Deserialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub enum CpuRequest {
+    Compact(NumOfCpus),
+    ForceCompact(NumOfCpus),
+    Scatter(NumOfCpus),
+    All,
+}
+
+impl Default for CpuRequest {
+    fn default() -> Self {
+        CpuRequest::Compact(1)
+    }
+}
+
+impl CpuRequest {
+    pub fn validate(&self) -> crate::Result<()> {
+        match &self {
+            CpuRequest::Scatter(n_cpus)
+            | CpuRequest::ForceCompact(n_cpus)
+            | CpuRequest::Compact(n_cpus) => {
+                if *n_cpus == 0 {
+                    Err(DsError::GenericError(
+                        "Zero cpus cannot be requested".to_string(),
+                    ))
+                } else {
+                    Ok(())
+                }
+            }
+            CpuRequest::All => Ok(()),
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub struct GenericResourceRequest {
+    pub resource: GenericResourceId,
+    pub amount: GenericResourceAmount,
+}
+
+pub type GenericResourceRequests = SmallVec<[GenericResourceRequest; 2]>;
+pub type TimeRequest = Duration;
+
+#[derive(Default, Serialize, Deserialize, Debug, Clone, Hash, Eq, PartialEq)]
+pub struct ResourceRequest {
+    cpus: CpuRequest,
+
+    // After normalization, this array is sorted by resource id
+    #[serde(default)]
+    generic: GenericResourceRequests,
+
+    /// Minimal remaining time of the worker life time needed to START the task
+    /// !!! Do not confuse with time_limit.
+    /// If task is started and task is running, it is not stopped if
+    /// it consumes more. If you need this, see time_limit in task configuration
+    /// On worker with not defined life time, this resource is always satisfied.
+    #[serde(default)]
+    min_time: TimeRequest,
+}
+
+impl ResourceRequest {
+    pub fn new(
+        cpu_request: CpuRequest,
+        time: TimeRequest,
+        generic_resources: GenericResourceRequests,
+    ) -> ResourceRequest {
+        ResourceRequest {
```

Comment:
It seems like this struct is mutable only because of tests (does it have to be?), but the `normalize` method is not test only. Shouldn't it be normalized in the constructor to make sure that it's in a valid state?

Reply:
It is mutated even in the normal code, as generic resources are added one-by-one. But it should be changed as you say